proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,507 @@
+`create` verb will be supported (but this resource will be part of the `pinniped` category and will have no-op `list`
+implementation just like token credential request to prevent `kubectl get pinniped -A` from returning an error).
+
+```go
+type OIDCClientSecretRequest struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`  // metadata.name must be set to the client ID
+
+
+  Spec   OIDCClientSecretRequestSpec   `json:"spec"`
+  Status OIDCClientSecretRequestStatus `json:"status"`
+}
+
+type OIDCClientSecretRequestSpec struct {
+  GenerateNewSecret bool `json:"generateNewSecret"`
+  RevokeOldSecrets  bool `json:"revokeOldSecrets"`
+}
+
+type OIDCClientSecretRequestStatus struct {
+  GeneratedSecret    string `json:"generatedSecret,omitempty"`
+  TotalClientSecrets int    `json:"totalClientSecrets"`
+}
+```
+
+Unlike token credential request, OIDC client secret request will require that `metadata.name` be set (so that it can
+determine what OAuth client is being referred to).  When `.spec.generateNewSecret` is set to `true`, the response will
+provide the plaintext client secret via the `.status.generatedSecret` field.  This is the only time that the plaintext
+client secret is made available.  To aid in rotation, this API may be called multiple times with
+`.spec.generateNewSecret` set to `true` to cause the creation of a new client secret.  The response will include the
+total number of client secrets (including any newly generated ones) that exist for the OAuth client in the
+`.status.totalClientSecrets` field.  When the admin is ready, they may call the API with `.spec.revokeOldSecrets` set to
+`true` to cause all but the latest secret to be revoked.  In the event of a client secret disclosure, a "hard" rotation
+may be performed by setting both `.spec.generateNewSecret` and `.spec.revokeOldSecrets` to `true` (this will revoke all
+pre-existing client secrets and return a newly generated secret).  Leaving both of these fields set to `false` will
+simply return the number of existing client secrets via the `.status.totalClientSecrets` field (this same information is
+available via the `.status.totalClientSecrets` field of the `OIDCClient` resource).
+
+An admin would interact with this API by using standard `kubectl` commands:
+
+```yaml
+apiVersion: oauth.virtual.supervisor.pinniped.dev/v1alpha1  # different group to avoid collision with the CRD
+kind: OIDCClientSecretRequest
+metadata:
+  name: client.oauth.pinniped.dev-j77kz
+  namespace: pinniped-supervisor
+spec:
+  generateNewSecret: true
+```
+
+Assuming the above yaml is stored in `file.yaml`, then running:
+
+`kubectl create -f file.yaml`
+
+would cause the server to respond with (note the custom columns in the table output):
+
+```
+NAMESPACE               NAME                                SECRET     TOTAL
+pinniped-supervisor     client.oauth.pinniped.dev-j77kz     12..xz     2
+```
```

Comment:
Won't I then miss the generated secret in the response status here unless I requested `-o yaml` or similar?

Reply:
@enj ?

Reply:
Related thought (based on above): iff the generated secret would usually end up being manually added as a secret resource in the users namespace, why not have the controller create the secret resource for the user rather than relying on catching the status in the first response. I think the same semantics for revocation etc. could still apply. It just depends whether the assumption can be made that a secret in the user's (requested?) namespace is secure.

Reply:
The user needs the plaintext secret so they can put that in their webapp's client config. @enj wanted to avoid storing the client secret in plain text anywhere, and also wanted the Supervisor to auto-generate the client secret to ensure that it is a strong (long, random) password. These two requirements together imply that the Supervisor needs to return the plaintext password one time upon creation, and then store the hashed version. Unfortunately, these requirements add a lot of complexity to both the design and implementation, but is the direction that we're going at the moment.

Reply:
> The user needs the plaintext secret so they can put that in their webapp's client config.

Yep, I guess I'm just asking if that client config will always end up being a combination of a k8s config map and k8s secret anyway (ie. will the user be creating a k8s secret with the generated client-secret to use in their webapp's config). If not, then I totally agree with the first requirement (as well as the second of course).

Reply:
There is no requirement that the app using the dynamic client be hosted on the same cluster as the supervisor (and I think we would prefer if it was not).  Said app may not even be running on Kubernetes.  Even when running on Kube, https://secrets-store-csi-driver.sigs.k8s.io can be used to store the client secret in an external vault and mounted into the app's pod without using Kubernetes secrets (which are really bad at actually being secret).

Reply:
> Won't I then miss the generated secret in the response status here unless I requested `-o yaml` or similar?

Sadly yes.  Even though all the code exists in the API server to return a proper table output on `create` (implemented in #1235), `kubectl` does not request it 😭